masq_lib/src/percentage.rs

```diff
@@ -80,178 +229,418 @@ impl Percentage {
+        assert_integer_compatibility(&subject, u16::MAX, 32768);
+        assert_integer_compatibility(&subject, u32::MAX, 2147483648);
+        assert_integer_compatibility(&subject, u64::MAX, 9223372036854775808);
+        assert_integer_compatibility(&subject, u128::MAX, 170141183460469231731687303715884105728);
+        assert_integer_compatibility(&subject, i8::MIN, -64);
+        assert_integer_compatibility(&subject, i16::MIN, -16384);
+        assert_integer_compatibility(&subject, i32::MIN, -1073741824);
+        assert_integer_compatibility(&subject, i64::MIN, -4611686018427387904);
+        assert_integer_compatibility(&subject, i128::MIN, -85070591730234615865843651857942052864);
+    }
+
+    fn assert_integer_compatibility<N>(subject: &PurePercentage, num: N, expected: N)
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+        i16: TryFrom<N>,
+        <i16 as TryFrom<N>>::Error: Debug,
+    {
+        assert_eq!(subject.of(num), expected);
+        let half = num / N::try_from(2).unwrap();
+        let one = N::try_from(1).unwrap();
+        assert!((half - one) <= half && half <= (half + one))
+    }
 
     #[test]
-    fn zero() {
-        assert_eq!(Percentage::new(45).of(0), 0);
-        assert_eq!(Percentage::new(0).of(33), 0)
+    fn zeros_for_pure_percentage() {
+        assert_eq!(PurePercentage::try_from(45).unwrap().of(0), 0);
+        assert_eq!(PurePercentage::try_from(0).unwrap().of(33), 0)
+    }
+
+    #[test]
+    fn pure_percentage_end_to_end_test_for_unsigned() {
+        let expected_values = (0..=100).collect::<Vec<i8>>();
+
+        test_end_to_end(100, expected_values, |percent, base| {
+            PurePercentage::try_from(percent).unwrap().of(base)
+        })
     }
 
     #[test]
-    fn end_to_end_test() {
-        let range = 0..=100;
+    fn pure_percentage_end_to_end_test_for_signed() {
+        let expected_values = (-100..=0).rev().collect::<Vec<i8>>();
+
+        test_end_to_end(-100, expected_values, |percent, base| {
+            PurePercentage::try_from(percent).unwrap().of(base)
+        })
+    }
+
+    fn test_end_to_end<F>(
+        base: i8,
+        expected_values: Vec<i8>,
+        create_percentage_and_apply_it_on_number: F,
+    ) where
+        F: Fn(u8, i8) -> i8,
+    {
+        let range = 0_u8..=100;
 
         let round_returned_range = range
-            .clone()
             .into_iter()
-            .map(|per_cent| Percentage::new(per_cent).of(100_u64))
-            .collect::<Vec<u64>>();
+            .map(|percent| create_percentage_and_apply_it_on_number(percent, base))
+            .collect::<Vec<i8>>();
 
-        let expected = range
-            .into_iter()
-            .map(|num| num as u64)
-            .collect::<Vec<u64>>();
-        assert_eq!(round_returned_range, expected)
+        assert_eq!(round_returned_range, expected_values)
     }
 
     #[test]
     fn only_numbers_up_to_100_are_accepted() {
```

Comment:
You could rename `PurePercentage` to `Percentage`. It was your idea, and I am convinced it's good to keep the name like this. Once you modify that, let's incorporate it in the name.